README.md
# Introduction

* iRedAPD is a simple [Postfix policy server](http://www.postfix.org/SMTPD_POLICY_README.html),
  written in Python and runs as a low-privileged user (`iredapd` by default),
  with plugin support.
* The latest iRedAPD works with OpenLDAP, MySQL/MariaDB and PostgreSQL backends.
* __License: GPL v3__, except few files are in different licenses, shipped in
  iRedAPD for easier setup, but not license change.
    * file `libs/daemon.py` is BSD license.
    * file `libs/srslib.py` is Apache License, Version 2.0.
* Author: Zhang Huangbin <zhb _at_ iredmail.org>.

**NOTES**:

* iRedAPD is a sub-project of [iRedMail project](http://www.iredmail.org).
* iRedAPD is installed and enabled in iRedMail by default, you don’t need
  this tutorial if you already have iRedMail running. For standard installation
  please check document `INSTALL.md`.
* iRedAPD listens on 3 ports by default:
    - `7777`: normal smtp policy service
    - `7778`: SRS (Sender Rewriting Scheme) for sender address rewriting
    - `7779`: SRS (Sender Rewriting Scheme) for recipient address rewriting
* You can manage iRedAPD with iRedMail [web admin panel - iRedAdmin-Pro](http://www.iredmail.org/admin_panel.html).

# Requirements

- Python 3.5+

# Manage iRedAPD with command line tools

iRedMail project has a detailed tutorial to show you how to manage iRedAPD
with command line tools: [Manage iRedAPD](http://www.iredmail.org/docs/manage.iredapd.html)

# Available plugins

Plugins are files placed under `plugins/` directory, plugin name is file name
without file extension `.py`. It's recommended to read comment lines in plugin
source files to understand what it does and how it works.

## Plugins for all backends

* `reject_to_hostname`: reject emails sent to `xxx@<server hostname>` from
  external network.
* `reject_sender_login_mismatch`: Reject sender login mismatch (addresses in
  `From:` and SASL username). It will verify user alias addresses against
  SQL/LDAP database.

    This plugin also verifies forged sender address, e.g. sending email as
    a local domain to local domain.

* `reject_null_sender`: Reject message submitted by sasl authenticated user but
  use null sender in `From:` header (`from=<>` in Postfix log).
  RECOMMENDED to enable this plugin. It doesn't require SQL/LDAP query.

    If your user's password was cracked by spammer, spammer can use
    this account to bypass smtp authentication, but with a null sender
    in `From:` header, throttling won't be triggered.

* `amavisd_wblist`: Whitelist/blacklist for both inbound and outbound messages.

    The white/blacklists are used by both iRedAPD (before-queue) and Amavisd
    (after-queue).

* `greylisting`: for greylisting service.
* `throttle`: Throttling based on:
    * max number of mail messages sent/received in specified period of time
    * total mail size sent in specified period of time
    * size of single message

* `whitelist_outbound_recipient`: automatically whitelist recipient addresses
  of outgoing emails sent by sasl authenticated (local) users. It's able to
  whitelist single recipient address or domain for greylisting and normal
  white/blacklist.

## Plugins for OpenLDAP backend

* `ldap_maillist_access_policy`: restrict who can send email to mail list.
* `ldap_force_change_password_in_days`: force users to change password in days (default 90 days). User cannot send email before resetting password.

## Plugins for MySQL/MariaDB and PostgreSQL backends

* `sql_alias_access_policy`: restrict who can send email to mail alias.
* `sql_force_change_password_in_days`: force users to change password in days (default 90 days). User cannot send email before resetting password.


==========================================

## Install on Ubuntu 20

* ENV

```sh
Python 3.8.10
OS: Ubuntu 20.04.3
```

* Install 

```sh
apt install python3-mysqldb -y
adduser --shell /sbin/nologin --home /home/iredapd --disabled-login iredapd

cd /opt
git clone -b wblist_extra https://github.com/iredmail/iRedAPD
ln -s /opt/iRedAPD /opt/iredapd
cd iredapd
cp rc_scripts/iredapd.debian /etc/init.d/iredapd

# Config log
mkdir /var/log/iredapd
touch /var/log/iredapd/iredapd.log
chmod +w /var/log/iredapd/iredapd.log
cp samples/rsyslog.d/iredapd.conf /etc/rsyslog.d/
cp samples/logrotate.d/iredapd /etc/logrotate.d/

systemctl restart rsyslog
systemctl restart logrotate.timer


# Config iredapd
cp  settings.py.sample  settings.py
pip3 install -r  requirements.txt

# Edit settings.py 
vim settings.py 

# Restart service 
systemctl daemon-reload
systemctl restart iredapd
systemctl enable iredapd

```





ChangeLog
Update by trangnth: 13/05/2021 (From branch old-version)
    * Limit max recipients in single message
    * Đếm số message được gửi đi, trong đó có email gửi ra ngoài 
    (tức là không quan tâm có bao nhiều người nhận là external, chỉ tính người gửi, gửi đi 1 message)

    * Before: 
        * Đếm số recipients external trong một message, cộng tổng lại theo từng user và từng domain


iRedAPD-2.2:
    * Improvements:
        + plugins/reject_sender_login_mismatch.py: If login mismatch (From: vs
          SASL username), bypass sender if sender server is listed in DNS SPF
          record.

          This is disabled by default, set `CHECK_SPF_IF_LOGIN_MISMATCH = True`
          in iRedAPD config file and restart iredapd service to enable it.

          Sponsored development by Kacper Guzik <kacper.guzik _at_ zenbox.pl>.

    * Fixed issues:
        - Not verify ip address/network specified in DNS SPF record with
          ip4/+ip4/ip6/+ip6 tag.
        - tools/upgrade_iredapd.sh: syntax error while creating 'wblist_rdns'
          SQL table with PostgreSQL command line.
          Thanks Nameless <nonsenseisay _at_ protonmail.com> for the report in
          forum.
        - plugins/ldap_maillist_access_policy.py: didn't handle deprecated
          policy name 'allowedonly'. it's still used by old iRedMail.
        - plugins/throttle.py: didn't apply throttle setting if recipient is an
          alias domain.
        - plugins/amavisd_wblist.py: didn't apply wblist if recipient is an
          alias domain.
          Thanks tom cotton <tom _at_ ansspc.com> for the report in forum.
        - Not escape dangerous characters in SQL queries.
        - iRedAPD was not ran as 'iredapd' group.
        - Don't return disabled domain as a local domain.
        - If client is sending email from trusted networks without SMTP AUTH,
          it's not considered as an internal sender and no outbound throttling
          applied.
          Thanks Ahmad M. Al-Masry <Ahmad.AlMasry _at_ iah.ps> for the report.

iRedAPD-2.1:
    * New plugins:
        + plugins/wblist_rdns.py: whitelisting and blacklisting based on
          reverse DNS name of sender server IP address.
          Sponsored development by Daniel Senie <dts _at_ amaranth.com>.

    * Improvements:
        + tools/spf_to_greylist_whitelists.py: clean up 'greylisting_tracking'
          table if certain ip is whitelisted.
        + plugins/amavisd_wblist.py: able to use CIDR as white/blacklists. e.g.
          192.168.1.0/24, 2002::1234:abcd:ffff:c0a8:101/64.
          Sponsored development by Daniel Senie <dts _at_ amaranth.com>.
        + plugins/greylisting.py:
            - if client passes, remove other tracking records from same IP
              immediately to keep database small.
            - if sender server already passed, renews the expire time when a
              new email from same sender server.
              Thanks Ian Knight <ian _at_ 34sp _dot_ com> for the code
              contrubtion.

    * Fixed issues:
        - '#user@domain.com' is not treated as valid email address.
          Thanks Ben Suffolk <ben _at_ vanilla.net> for the report.
        - Not strip address extension in sender/recipient mail address.
        - plugins/greylisting.py: not check sub-domain settings.
          Thanks mptg <info _at_ marcingolanski.pl> for the report in forum.
        - plugins/throttle.py: restriction rule based on '@ip' doesn't work.
          Thanks raz <raz _at_ dtnt.com> for the report in forum.
        - plugins/reject_sender_login_mismatch.py:
            - should not query alias domain when sender domain is same as
              smtp auth username domain.
            - not check per-user alias addresses.
              Thanks Wraptor <nijs.thibaut _at_ gmail> for the report in forum.
            - should not consider backup mx domain name as local domain.
              Thanks Znekar <znekar _at_ znekar.com> for the report in forum.
            - [ldap] not check domain status while getting account ldif data.
              Thanks Ben Lin <ben_lin _at_ opl.com.tw> for the report.

iRedAPD-2.0:
    * New plugins:
        - plugins/whitelist_outbound_recipient.py: used to automatically
          whitelist outbound recipient for greylisting (parameter
          'WL_RCPT_FOR_GREYLISTING'). Please check comment lines in
          `libs/default_settings.py` for more details.

    * Improvements:
        - Copy `plugins/custom_*` during upgrading to keep custom plugins.
        - Able to throttle based on external senders.
        - Handle systemd script while upgrading in tools/upgrade_iredapd.sh.
        - Return 'DISCARD' instead of 'REJECT' to Postfix so that blacklisted
          sender won't know it's blacklisted.
        - If one sender passed greylisting service, we reuse the sender IP
          address to avoid further temporary rejection.
        - Get full traceback if error occurs while applying plugin.
        - Reduce SQLAlchemy connection pool recycle to 360 (was 3600) to work
          better on low traffic server.
        - SQLAlchemy connection pool size and pool recycle are configurable
          with parameter `SQL_CONNECTION_POOL_SIZE` and
          `SQL_CONNECTION_POOL_RECYCLE` in iRedAPD config file.
        - tools/greylisting_admin.py::
            + able to specify IP address/network, email address as sender.
            + able to add/list per-account whitelists.
        - Add systemd script used to control iredapd service:
          rc_scripts/iredapd.service.

    * Fixed issues:
        - Memory leak due to not clean up session tracking data.
        - plugins/throttle.py: Throttle plugin cannot handle email which has
          multiple recipients.
          Thanks to Andrea <info _at_ nili.ca> for the report.
        - plugins/amavisd_wblist.py: not apply inbound wblist if both sender
          and recipient are local virtual mail user.
          Thanks Tom <tom _at_ ansspc.com> for the report in forum.

libs/__init__.py
__author__ = 'Zhang Huangbin <zhb@iredmail.org>'
__version__ = '2.2'


SMTP_ACTIONS = {
    'default': 'DUNNO',
    # Use 'OK' carefully, it will bypass other Postfix/iRedAPD restrictions.
    'accept': 'OK',
    'whitelist': 'OK',
    # discard email without return error message to sender
    'discard': 'DISCARD Policy discard',
    # reject
    'reject': 'REJECT Policy rejection',
    # reject with reason
    'reject_null_sender': 'REJECT Policy rejection due to null sender',
    'reject_forged_sender': 'REJECT SMTP AUTH is required for users under this sender domain',
    'reject_sender_login_mismatch': 'REJECT Sender is not same as SMTP authenticate username',
    'reject_blacklisted': 'REJECT Blacklisted',
    'reject_not_authorized': 'REJECT Not authorized',
    'reject_message_size_exceeded': 'REJECT Message size exceed (maybe caused by big attachment file)',
    'reject_blacklisted_rdns': 'REJECT Blacklisted reverse DNS name of server IP address',
    # Throttling
    'reject_exceed_msg_size': 'REJECT Quota exceeded (size of single mail message)',
    'reject_exceed_max_msgs': 'REJECT Quota exceeded (number of mails in total)',
    'reject_exceed_max_quota': 'REJECT Quota exceeded (accumulated message size)',
    'reject_exceed_max_rcpts': 'REJECT Too many recipients in single message',
    'greylisting': '451 4.7.1',
}

# Default replies for Postfix tcp table.
# Reference: http://www.postfix.org/tcp_table.5.html
TCP_REPLIES = {
    # In case of a lookup request, the requested data does not exist.
    # In case of an update request, the request was rejected. The text
    # describes the nature of the problem.
    'default': '500 No result',

    # This indicates an error condition. The text describes the nature
    # of the problem. The client should retry the request later.
    'error': '400 Error ',

    # The request was successful. In the case of a lookup request,
    # the text contains an encoded version of the requested data.
    'found': '200 ',
}

# Plugin priorities.
#
# * With pre-defined priorities, the order defined in `plugins = []` setting
#   doesn't matter at all, so that we can apply plugins in ideal order.
#
# * It's better to run plugins which doesn't require SQL/LDAP connection first.
#               +-----------------------------------+
# * Plugin with | larger number has higher priority | and will be applied first.
#               +-----------------------------------+
PLUGIN_PRIORITIES = {
    'reject_null_sender': 100,
    'reject_to_hostname': 100,
    'wblist_rdns': 99,
    'reject_sender_login_mismatch': 90,
    'greylisting': 80,
    'ldap_force_change_password_in_days': 70,
    'sql_force_change_password_in_days': 70,
    'throttle': 60,
    'ldap_maillist_access_policy': 50,
    'sql_alias_access_policy': 50,
    'amavisd_wblist': 40,
    'whitelist_outbound_recipient': 10,
}

# Account proiroties.
# Used in plugins:
#   - greylisting.py
ACCOUNT_PRIORITIES = {
    'email': 100,               # e.g. 'user@domain.com'. Highest priority
    'wildcard_addr': 90,        # e.g. `user@*`. used in plugin `amavisd_wblist`
                                # as wildcard sender. e.g. 'user@*`
    'ip': 80,                   # e.g. 173.254.22.21
    'wildcard_ip': 70,          # e.g. 173.254.22.*
    'cidr': 70,                 # e.g. 173.254.22.0/24
    'domain': 60,               # e.g. @domain.com
    'subdomain': 50,            # e.g. @.domain.com
    'top_level_domain': 40,     # e.g. @com, @org
    'catchall': 0,              # '@.'. Lowest priority
}

# Mail list access policies.
# Unrestricted
MAILLIST_POLICY_PUBLIC = 'public'
# Only users under same domain are allowed
MAILLIST_POLICY_DOMAIN = 'domain'
# Only users under same domain and sub domains are allowed
MAILLIST_POLICY_SUBDOMAIN = 'subdomain'
# Only members are allowed
MAILLIST_POLICY_MEMBERSONLY = 'membersonly'
# Only moderators/allowed are allowed
MAILLIST_POLICY_MODERATORS = 'moderatorsonly'
# Only members and moderators are allowed
MAILLIST_POLICY_MEMBERSANDMODERATORSONLY = 'membersandmoderatorsonly'


# All the attributes that the Postfix SMTP server sends in a delegated SMTPD
# access policy request.
# Reference: http://www.postfix.org/SMTPD_POLICY_README.html
SMTP_SESSION_ATTRIBUTES = [
    # Postfix version 2.1 and later:
    'request',
    'protocol_state',
    'protocol_name',
    'helo_name',
    'queue_id',             # Empty in RCPT state
    'sender',
    'recipient',
    'recipient_count',      # Empty in RCPT state
    'client_address',
    'client_name',
    'reverse_client_name',
    'instance',
    # Postfix version 2.2 and later:
    'sasl_method',
    'sasl_username',
    'sasl_sender',
    'size',                 # Empty in RCPT
    'ccert_subject',
    'ccert_issuer',
    'ccert_fingerprint',
    # Postfix version 2.3 and later:
    'encryption_protocol',
    'encryption_cipher',
    'encryption_keysize',
    'etrn_domain',
    # Postfix version 2.5 and later:
    'stress',
    # Postfix version 2.9 and later:
    'ccert_pubkey_fingerprint',
    # Postfix version 3.0 and later:
    'client_port',
    # Postfix version 3.1 and later:
    'policy_context',
    # Postfix version 3.2 and later:
    'server_address',
    'server_port',
]
